tools/cli/commands/generate.js

```diff
@@ -0,0 +1,236 @@
+/**
+ * External dependencies
+ */
+import path from 'path';
+import fs from 'fs';
+import pluralize from 'pluralize';
+import inquirer from 'inquirer';
+
+/**
+ * Internal dependencies
+ */
+import { promptForType, promptForName } from '../helpers/promptForProject.js';
+import { projectTypes, checkNameValid } from '../helpers/projectHelpers.js';
+import {
+	readPackageJson,
+	readComposerJson,
+	writePackageJson,
+	writeComposerJson,
+} from '../helpers/json';
+import { normalizeGenerateArgv } from '../helpers/normalizeArgv';
+import mergeDirs from '../helpers/mergeDirs';
+
+/**
+ * Relays commands to generate a particular project
+ *
+ * @param {object} options - The argv options.
+ */
+async function generateRouter( options ) {
+	const argv = normalizeGenerateArgv( options );
+	generatePackage( argv );
+	console.log( argv );
+
+	//Route the project to the correct function to be built here.
+}
+
+/**
+ * Entry point for the CLI.
+ *
+ * @param {object} argv - The argv for the command line.
+ */
+export async function generateCli( argv ) {
+	argv = await promptForGenerate( argv );
+	await generateRouter( argv );
+}
+
+/**
+ * Command definition for the generate subcommand.
+ *
+ * @param {object} yargs - The Yargs dependency.
+ *
+ * @returns {object} Yargs with the generate commands defined.
+ */
+export function generateDefine( yargs ) {
+	yargs.command(
+		'generate [type]',
+		'Creates a new project',
+		yarg => {
+			yarg
+				.positional( 'type', {
+					describe: 'Type of project being worked on, e.g. package, plugin, etc',
+					type: 'string',
+				} )
+				.options( 'name', {
+					alias: 'n',
+					describe: 'Name of the project',
+					type: 'string',
+				} );
+		},
+		async argv => {
+			await generateCli( argv );
+			if ( argv.v ) {
+				console.log( argv );
+			}
+		}
+	);
+
+	return yargs;
+}
+
+/**
+ * Prompt for Generating New Project.
+ *
+ * If no project is passed via `options`, then it will prompt for the type of project and the project itself.
+ *
+ * @param {object} options - Passthrough of the argv object.
+ *
+ * @returns {object} argv object with the project property.
+ */
+export async function promptForGenerate( options ) {
+	let typeAnswer = options.type ? { type: options.type } : '';
+	let nameAnswer;
+	let tries = 0;
+
+	// Get project type if not passed as an option.
+	if ( ! typeAnswer || typeAnswer.length === 0 ) {
+		typeAnswer = await promptForType();
+	} else if ( ! projectTypes.includes( pluralize( typeAnswer.type ) ) ) {
+		return new Error( 'Must be a valid project type' );
```

Comment:
```suggestion
		throw new Error( 'Must be a valid project type' );
```

Reply:
We should throw errors instead of returning them. The next try/catch that exists is the highest level one in the bin right now, but we can improve that too.